libraries/abstractions/common_io/test/test_iot_spi.c

```diff
@@ -1479,7 +1486,221 @@ TEST( TEST_IOT_SPI, AFQP_IotSPI_CancelFuzzing )
+    TEST_ASSERT_NOT_EQUAL( NULL, xSPIHandle );
+
+    if( TEST_PROTECT() )
+    {
+        /* Adjust the dummy value */
+        TEST_ASSERT( IOT_SPI_SUCCESS == iot_spi_ioctl( xSPIHandle, eSPIGetMasterConfig, &xConfig ) );
+        xConfig.ucDummyValue = 0xDC;
+        TEST_ASSERT( IOT_SPI_SUCCESS == iot_spi_ioctl( xSPIHandle, eSPISetMasterConfig, &xConfig ) );
+
+        /* Install callback which should not be exercised because it's sync */
+        iot_spi_set_callback( xSPIHandle, prvSpiAsyncCallback, NULL );
+
+        /* The callback should not be called to give sem back. Read, verify callback not called */
+        TEST_ASSERT( IOT_SPI_SUCCESS == iot_spi_read_sync( xSPIHandle, pucRxBuf, ulNBytes ) );
+        TEST_ASSERT( pdFAIL == xSemaphoreTake( xtestIotSPISemaphore, testIotSPI_DEFAULT_SEMAPHORE_DELAY / 2 ) );
+
+        /* Verify expected bytes. Because it's loopback, the configured dummy value
+         * should be transmitted, and thus read, for all N queried bytes.*/
+        TEST_ASSERT_EACH_EQUAL_UINT8( xConfig.ucDummyValue, pucRxBuf, ulNBytes );
+    }
+
+    lRetVal = iot_spi_close( xSPIHandle );
+    TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Assumes HW loopback where MOSI --> MISO.
+ *        Verify the happy-path of a async read
+ *
+ */
+TEST( TEST_IOT_SPI, IotSPI_LoopBack_ReadAsync )
+{
+    IotSPIHandle_t xSPIHandle;
+    int32_t lRetVal;
+    IotSPIMasterConfig_t xConfig;
+    const uint32_t ulNBytes = 16;
+    uint8_t pucRxBuf[ ulNBytes ];
+
+    memset( pucRxBuf, 0, ulNBytes );
+
+    xSPIHandle = iot_spi_open( ultestIotSpiInstance );
+    TEST_ASSERT_NOT_EQUAL( NULL, xSPIHandle );
+
+    if( TEST_PROTECT() )
+    {
+        /* Adjust the dummy value */
+        TEST_ASSERT( IOT_SPI_SUCCESS == iot_spi_ioctl( xSPIHandle, eSPIGetMasterConfig, &xConfig ) );
+        xConfig.ucDummyValue = 0xDC;
+        TEST_ASSERT( IOT_SPI_SUCCESS == iot_spi_ioctl( xSPIHandle, eSPISetMasterConfig, &xConfig ) );
+
+        iot_spi_set_callback( xSPIHandle, prvSpiAsyncCallback, NULL );
+
+        /* Read, then verify callback is called */
+        TEST_ASSERT( IOT_SPI_SUCCESS == iot_spi_read_async( xSPIHandle, pucRxBuf, ulNBytes ) );
+        TEST_ASSERT( pdPASS == xSemaphoreTake( xtestIotSPISemaphore, testIotSPI_DEFAULT_SEMAPHORE_DELAY ) );
+
+        /* Verify expected bytes. Because it's loopback, the configured dummy value
+         * should be transmitted, and thus read, for all N queried bytes.*/
+        TEST_ASSERT_EACH_EQUAL_UINT8( xConfig.ucDummyValue, pucRxBuf, ulNBytes );
+    }
+
+    lRetVal = iot_spi_close( xSPIHandle );
+    TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
+}
+
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Assumes HW loopback where MOSI --> MISO.
+ *        Verify the happy-path of a syncronous write
+ *
+ */
+TEST( TEST_IOT_SPI, IotSPI_LoopBack_WriteSync )
```

Comment:
I see there is no verification in these two "write" tests, compared to read or transfer loopback tests. If it is difficult to add verification code, I would say remove these two test cases to not provide a wrong impression.

Reply:
The main check in these tests is to make sure that after sync the user callback is NOT called, and for async the user callback IS called. Exercised by:
`TEST_ASSERT( pdFAIL == xSemaphoreTake( xtestIotSPISemaphore, testIotSPI_DEFAULT_SEMAPHORE_DELAY / 2 ) );`

Can't verify the MISO data as it's a write as there's no RxBuffer and MISO is ignored/undefined.

Reply:
ok, maybe you can call read_async before calling the write? Then you can verify the the data returned by read_async.

Reply:
This could likely work with separate Master and Slave instances, but currently the CommonIO only supports Master perspective. 

Wouldn't be possible using the same Master instance since async/sync transfers can't occur at the same time. A Master read is still a transfer, just one that only transmits dummy bytes while it clocks in MISO. Any read transaction that occurs with MOSI --> MISO tied will only read the dummy byte.

Couldn't do it with two masters connected to each other since data synchronization is needed and both will attempt to drive SCLK which is prohibited.